cpp/oneapi/dal/exceptions.hpp

```diff
@@ -82,9 +82,17 @@ class ONEAPI_DAL_EXPORT internal_error : public runtime_error, public std::runti
     const char* what() const noexcept override;
 };
 
-class ONEAPI_DAL_EXPORT bad_alloc : public exception, public std::bad_alloc {
+class ONEAPI_DAL_EXPORT bad_alloc : public exception, public std::bad_alloc {};
+
+class ONEAPI_DAL_EXPORT host_bad_alloc : public bad_alloc {
+public:
+    host_bad_alloc() noexcept = default;
+    const char* what() const noexcept override;
+};
+
+class ONEAPI_DAL_EXPORT device_bad_alloc : public bad_alloc {
```

Comment:
Don't we want to add information about which device cause the exception?

Reply:
My belief - no. `std::bad_alloc` provides only default message to save the memory. So, if we see strong need, new sub-classes like `dal::gpu_bad_alloc` will be implemented.

Reply:
please, resolve if agree

Reply:
Seems very strange for me to have an exception class for each device type like gpu, fpga. etc. What if I have two GPUs? Which device cases exception? How to know that? 
Have you looked at how `device_bad_alloc` is organized in other oneAPI libraries?

Reply:
decided to postpone.